hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/MarkerHandler.java

```diff
@@ -0,0 +1,297 @@
+                       int batchNumThreads, long batchIntervalMs, int parallelism) throws IOException {
+    super(conf, fileSystem, viewManager);
+    LOG.info("*** MarkerHandler FileSystem: " + this.fileSystem.getScheme());
+    LOG.info("*** MarkerHandler Params: batchNumThreads=" + batchNumThreads + " batchIntervalMs=" + batchIntervalMs + "ms");
+    this.metricsRegistry = metricsRegistry;
+    this.batchIntervalMs = batchIntervalMs;
+    this.parallelism = parallelism;
+    this.executorService = Executors.newScheduledThreadPool(batchNumThreads);
+    List<Boolean> isMarkerFileInUseList = new ArrayList<>(batchNumThreads);
+    for (int i = 0; i < batchNumThreads; i++) {
+      isMarkerFileInUseList.add(false);
+    }
+    this.isMarkerFileInUseList = Collections.synchronizedList(isMarkerFileInUseList);
+  }
+
+  public Set<String> getAllMarkers(String markerDirPath) {
+    return allMarkersMap.getOrDefault(markerDirPath, new HashSet<>());
+  }
+
+  public Set<String> getCreateAndMergeMarkers(String markerDirPath) {
+    return allMarkersMap.getOrDefault(markerDirPath, new HashSet<>()).stream()
+        .filter(markerName -> !markerName.endsWith(IOType.APPEND.name()))
+        .collect(Collectors.toSet());
+  }
+
+  public CompletableFuture<String> createMarker(Context context, String markerDirPath, String markerName) {
+    LOG.info("Request: create marker " + markerDirPath + " " + markerName);
+    CreateMarkerCompletableFuture future = new CreateMarkerCompletableFuture(context, markerDirPath, markerName);
+    synchronized (createMarkerRequestlockObject) {
+      createMarkerFutures.add(future);
+      long currTimeMs = System.currentTimeMillis();
+      if (currTimeMs >= nextBatchProcessTimeMs - MARGIN_TIME_MS) {
+        nextBatchProcessTimeMs += batchIntervalMs * (Math.max((currTimeMs - nextBatchProcessTimeMs) / batchIntervalMs + 1L, 1L));
+
+        long waitMs = nextBatchProcessTimeMs - System.currentTimeMillis();
+        executorService.schedule(
+            new BatchCreateMarkerRunnable(), Math.max(0L, waitMs), TimeUnit.MILLISECONDS);
+        LOG.info("Wait for " + waitMs + " ms, next batch time: " + nextBatchProcessTimeMs);
+      }
+    }
+    return future;
+  }
+
+  public Boolean deleteMarkers(String markerDir) {
+    Path markerDirPath = new Path(markerDir);
+    try {
+      if (fileSystem.exists(markerDirPath)) {
+        FileStatus[] fileStatuses = fileSystem.listStatus(markerDirPath);
+        List<String> markerDirSubPaths = Arrays.stream(fileStatuses)
+            .map(fileStatus -> fileStatus.getPath().toString())
+            .collect(Collectors.toList());
+
+        if (markerDirSubPaths.size() > 0) {
+          for (String subPathStr: markerDirSubPaths) {
+            fileSystem.delete(new Path(subPathStr), true);
+          }
+        }
+
+        boolean result = fileSystem.delete(markerDirPath, true);
```

Comment:
I plan to implement the logic of concurrently deleting marker files / subpaths so that's why I'd like to get a list of subpaths first.  I'll add this after I address all the comments.

Reply:
I've updated the logic here.  Deletion of marker files in the directory is parallelized to be more efficient, and we need to get the sub-directories, i.e., marker file paths, first in the directory.